File: backend/opencheck/sources/eiti_assessment.py
"""EITI Company Assessment adapter — ESG category.

This is a **third** EITI product, distinct from the two OpenCheck already has.
``eiti`` targets the EITI API and surfaces company payments to governments;
``eiti_soe`` targets the state-owned-enterprise roster. This adapter targets the
**Company Assessment**, which arrived with the new EITI global database
(``eiti-database.eiti.org``, 2026) and was absent from the SOE database
entirely.

The Company Assessment evaluates EITI *supporting companies* — the ~99
BHP/Anglo/Shell-class multinationals that fund and endorse the EITI — against
nine expectations. Two of them are the reason this adapter exists:

* **Expectation 6 — "Company disclose beneficial ownership."** A per-company,
  per-year result, with links to the disclosure where EITI captured them. This
  is the closest thing to a beneficial ownership signal EITI publishes.
* **Expectation 2 — "Company publish a list of controlled subsidiaries."**
  Plus the declared list itself: 1,230 parent→child rows across 51 EITI
  implementing countries.

What this source is *not*
-------------------------
It is **not beneficial ownership data**. Expectation 6 records a company's
*disclosure posture* — whether it says it discloses, and where — not who owns
it. The distinction matters and is carried through every layer here: no risk
signal, no person statements, no ownership relationships. The advocacy point
this data supports is that these companies have committed to disclosing
beneficial ownership and overwhelmingly do it as a PDF in an annual report
rather than as structured data, which is a weaker form of transparency than the
commitment implies.

The declared subsidiaries are **names and countries only**. EITI publishes no
identifier for them, no percentage and no share class.

.. important::
   **The declared subsidiary names must never enter the BODS graph** — no
   entity statements, no relationship statements, no nodes, no edges. They are
   evidence rendered on a card. Matching 1,230 free-text names to companies is
   a different and much harder problem than the 99 parents, and asserting an
   unmatched name as a graph node would claim an identity OpenCheck has not
   established. ``map_eiti_assessment`` emits one entity statement for the
   subject and nothing else; ``tests/test_eiti_assessment.py`` pins that.

Offline by design
-----------------
Identity resolution happens **once, offline**, in
``scripts/build_eiti_assessment_index.py``, which resolves the supporting
companies to LEIs and refuses to write an index until a human has reviewed
every match. At runtime this adapter loads the committed, LEI-keyed artifact
and answers ``fetch_by_lei`` as a dict lookup — no network on the hot path, and
no live-enrichment path at all.

The review gate is not ceremony. A name matcher over multinational-enterprise
names produces confident nonsense: "Anglo American" ranks the ANGLO AMERICAN
FOUNDATION first, "Equinor" once matched a company sports club, and stripping
legal-form suffixes makes a group head exactly equal to its own subsidiaries
("Glencore" → GLENCORE AG). Review caught a match on TECK GmbH, an unrelated
German company, that had passed as an exact name match.

Identifier corroboration
------------------------
This adapter asserts **no identifiers at all**, which is deliberate:

* The **LEI is derived** by OpenCheck at build time, so the corroboration rule
  in ``CLAUDE.md`` bars asserting it.
* EITI's ``legal_entity_id`` column exists but is populated for **3 companies
  of 10,116**, so it is not a usable source of LEIs.
* EITI's ``eiti_id_company`` is a **UUIDv5 over a name-derived key** — a
  deduplication key that EITI regenerated wholesale (v4 → v5) when it launched
  this database, not a registry number. Publishing it as an identifier would
  invite consumers to store something EITI can legitimately change.

An empty identifier set is the honest outcome here, not an oversight.

Licence: EITI content-use policy — free republication with credit to
"EITI International Secretariat, eiti.org".
"""

from __future__ import annotations

import asyncio
import gzip
import json
import logging
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .. import provenance
from .base import SearchKind, SourceAdapter, SourceHit, SourceInfo

log = logging.getLogger(__name__)

#: Committed, LEI-keyed index (built by scripts/build_eiti_assessment_index.py).
#: Overridable via env for tests / alternative snapshots.
_INDEX_PATH = Path(
    os.environ.get("EITI_ASSESSMENT_INDEX_PATH", "")
    or (Path(__file__).resolve().parent.parent / "data" / "eiti_assessment_index.json.gz")
)

#: Lazy module-level singleton (LEI -> record). Tests may set this directly.
_index: dict[str, dict[str, Any]] | None = None
#: Upstream extract date of the committed index (``meta.source_harvest``) —
#: when EITI's data was actually read, not when git wrote the file here.
_source_snapshot: str | None = None

#: BODS shorthand for the expectation that carries the beneficial ownership
#: assessment. Named once so the frontend, the mapper and the findings template
#: cannot drift apart on a string literal.
BO_EXPECTATION = "exp_6"
#: The controlled-subsidiaries expectation.
SUBSIDIARY_EXPECTATION = "exp_2"


def _get_index() -> dict[str, dict[str, Any]]:
    """Load the committed LEI-keyed index (cached in a module singleton)."""
    global _index, _source_snapshot
    if _index is None:
        try:
            with gzip.open(_INDEX_PATH, "rt", encoding="utf-8") as f:
                data = json.load(f)
            raw = data.get("index") or {}
            _index = {
                str(k).strip().upper(): v
                for k, v in raw.items()
                if len(str(k).strip()) == 20
            }
            meta = data.get("meta") or {}
            _source_snapshot = meta.get("source_harvest") or meta.get("built")
            log.info(
                "EITI Company Assessment index loaded: %s of %s supporting "
                "companies resolved to an LEI, %s declared subsidiaries (%s harvest)",
                meta.get("parents_accepted", len(_index)),
                meta.get("parents_total"),
                meta.get("subsidiaries_indexed"),
                meta.get("source_harvest"),
            )
        except Exception as exc:  # noqa: BLE001
            log.warning("EITI Company Assessment index unavailable: %s", exc)
            _index = {}
    return _index


def _snapshot_dt() -> datetime | None:
    """The index's upstream harvest date as a datetime, or None."""
    if not _source_snapshot:
        return None
    try:
        dt = datetime.fromisoformat(str(_source_snapshot).replace("Z", "+00:00"))
    except ValueError:
        return None
    return dt if dt.tzinfo else dt.replace(tzinfo=timezone.utc)


def _load_and_declare() -> dict[str, dict[str, Any]]:
    """Load the index and declare the payload as curated, never live.

    Without this the adapter records nothing on a match and the bundle would
    resolve to 'stub', under-claiming: the data is real, it is simply a
    snapshot. The declared date is EITI's harvest date, not the file's mtime —
    an mtime only records when git wrote the file to this machine, which says
    nothing about how current the data is.
    """
    index = _get_index()
    provenance.record_curated(
        "EITI Company Assessment snapshot committed to the repository",
        harvested_at=_snapshot_dt(),
    )
    return index


def _reset_index_for_tests() -> None:
    """Test helper — drop the cached singleton so a fresh index is loaded."""
    global _index, _source_snapshot
    _index = None
    _source_snapshot = None


class EitiAssessmentAdapter(SourceAdapter):
    """EITI Company Assessment adapter — ESG category."""

    id = "eiti_assessment"

    #: LEI-keyed source, dispatched directly in routers/lookup.py alongside the
    #: other LEI-keyed offline sources (eiti_soe, eiti_bo, cac_nigeria), not via
    #: an RA-code deriver. No network on the hot path, so the budget is small.
    lookup_timeout_s = 10.0

    @property
    def info(self) -> SourceInfo:
        return SourceInfo(
            id=self.id,
            name="EITI Company Assessment",
            homepage="https://eiti-database.eiti.org/",
            description=(
                "EITI's assessment of its supporting companies against nine "
                "expectations, including whether the company discloses its "
                "beneficial owners and publishes a list of controlled "
                "subsidiaries — plus the declared subsidiary list itself. "
                "Records a company's disclosure posture, not its ownership. "
                "Snapshot of the EITI global database, resolved to LEIs "
                "offline and reviewed by hand."
            ),
            license="EITI content-use policy (free reuse with attribution)",
            attribution="EITI International Secretariat, eiti.org",
            supports=[SearchKind.ENTITY],
            requires_api_key=False,
            live_available=bool(_get_index()),
            category="esg",
            is_national_register=False,
            country=None,
        )

    async def search(self, query: str, kind: SearchKind) -> list[SourceHit]:
        # Identifier-keyed (LEI) source; free-text search intentionally empty.
        return []

    def covers_lei(self, lei: str) -> bool:
        """Whether the committed assessment snapshot holds this LEI.

        The lookup pipeline asks before dispatching, so a company this file
        cannot describe is never announced as a source being queried and never
        counted in "N of N sources answered". The set is EITI's supporting
        companies — a few dozen multinationals — not every extractive company,
        so this governs whether the source is *applicable* and says nothing
        about the company.

        Reads the index without declaring provenance: nothing has been fetched.
        """
        return (lei or "").strip().upper() in _get_index()

    async def fetch_by_lei(self, lei: str) -> dict[str, Any] | None:
        """Return the assessment bundle for a LEI, or ``None`` when absent."""
        lei_norm = (lei or "").strip().upper()
        index = await asyncio.to_thread(_load_and_declare)
        record = index.get(lei_norm)
        if record is None:
            return None
        return self._build_bundle(lei_norm, record)

    async def fetch(self, hit_id: str) -> dict[str, Any]:
        """Fetch by LEI hit id (deepen / retry path)."""
        lei_norm = (hit_id or "").strip().upper()
        index = await asyncio.to_thread(_load_and_declare)
        record = index.get(lei_norm)
        if record is None:
            return {"source_id": self.id, "hit_id": hit_id, "is_stub": True}
        return self._build_bundle(lei_norm, record)

    def _build_bundle(self, lei: str, record: dict[str, Any]) -> dict[str, Any]:
        # Corroboration rule: assert NOTHING. The LEI is OpenCheck-derived, and
        # neither identifier EITI publishes is usable as one — see the module
        # docstring. `eiti_published` below carries them for display only.
        return {
            "source_id": self.id,
            "hit_id": lei,
            "lei": lei,
            "is_stub": False,
            "identifiers": {},
            "name": record.get("name") or "",
            "hq_country": record.get("hq_country"),
            "hq_city": record.get("hq_city"),
            "sectors": record.get("sectors") or [],
            "company_type": record.get("company_type"),
            "business_activity": record.get("business_activity"),
            "eiti_published": record.get("eiti_published") or {},
            "match": record.get("match") or {},
            "assessments": record.get("assessments") or {},
            "subsidiaries": record.get("subsidiaries") or [],
            "source_snapshot": _source_snapshot,
        }


# --------------------------------------------------------------------------
# Reading helpers — shared by the hit builder, the mapper and the findings
# template so the three cannot disagree about what the data says.
# --------------------------------------------------------------------------


def latest_year(bundle: dict[str, Any]) -> str | None:
    """The most recent assessment year present, as a string, or None."""
    years = [y for y in (bundle.get("assessments") or {}) if str(y).strip()]
    return max(years, key=lambda y: (len(y), y)) if years else None


def expectation(bundle: dict[str, Any], shorthand: str, year: str | None = None):
    """One expectation's fields for *year* (default: the latest year)."""
    year = year or latest_year(bundle)
    if not year:
        return None
    return ((bundle.get("assessments") or {}).get(year) or {}).get(shorthand)


def bo_disclosure_result(bundle: dict[str, Any], year: str | None = None) -> str | None:
    """EITI's verbatim expectation-6 result, e.g. "Expectation met".

    Returned as EITI wrote it. "Not available" means EITI did not assess, which
    is not a finding against the company, and the frontend must not render it
    in a risk tone — see ``RISK`` note in the ESG card.
    """
    exp = expectation(bundle, BO_EXPECTATION, year)
    return (exp or {}).get("result") or None


def bo_disclosure_url(bundle: dict[str, Any], year: str | None = None) -> str | None:
    """The best available link to the company's beneficial ownership disclosure.

    EITI populated these for the 2023 cohort and **not** for 2025, so a lookup
    on the latest year alone silently shows nothing for most companies. This
    falls back through earlier years rather than pretending no disclosure link
    exists; the caller should say which year the link came from.
    """
    years = sorted((bundle.get("assessments") or {}), reverse=True)
    if year:
        years = [year]
    for y in years:
        exp = ((bundle.get("assessments") or {}).get(y) or {}).get(BO_EXPECTATION) or {}
        for field in ("bo_url", "bo_disclosure_url", "url"):
            value = (exp.get(field) or "").strip()
            if value:
                return value
    return None

File: frontend/src/lib/eitiAssessment.ts
/**
 * The EITI Company Assessment, as values (Phase 11, plan B3).
 *
 * Everything on the `eiti_assessment` card that can be wrong about *what the
 * data says* lives here rather than in the markup, because the two claims this
 * source makes are both easy to overstate by accident:
 *
 * 1. **A disclosure assessment is not a finding about the company.** EITI's
 *    expectation 6 asks whether a supporting company discloses its beneficial
 *    owners. A result of "Not available" means EITI did not assess it — and
 *    rendering that in a risk tone would convert a gap in EITI's coverage into
 *    an allegation. `boDisclosureTone` is the one place that mapping exists,
 *    and `eitiAssessment.test.ts` pins the absent cases to non-risk tones.
 * 2. **A name in EITI's subsidiary list is not a matched company.** EITI
 *    publishes 1,230 parent→child rows as free text with no identifier. The
 *    cross-reference below compares *names* and says so; it never claims the
 *    two records are the same company.
 *
 * ## Built for more than one comparison list
 *
 * `crossReference` takes N lists, not one. Today the only list is the GLEIF
 * Level 2 children OpenCheck already fetches; B4 adds OECD-UNSD MEIP once its
 * BODS release lands, and that must be an added column rather than a rewrite of
 * this module and the card above it.
 */

import { compareKey } from "./vocab";

export const BO_EXPECTATION = "exp_6";
export const SUBSIDIARY_EXPECTATION = "exp_2";

// ---------------------------------------------------------------------------
// The bundle, as the adapter builds it
// ---------------------------------------------------------------------------

export interface EitiExpectation {
  label?: string | null;
  result?: string | null;
  response?: string | null;
  url?: string | null;
  comment?: string | null;
  bo_disclosure?: string | null;
  bo_url?: string | null;
  bo_disclosure_url?: string | null;
  stock_exchange?: string | null;
  stock_url?: string | null;
}

export interface EitiDeclaredSubsidiary {
  name: string;
  country?: string | null;
  years?: string[];
  /** EITI's own note on where the row came from ("Source: Company feedback"). */
  source?: string | null;
}

export interface EitiAssessmentMatch {
  method?: string | null;
  reviewed?: boolean;
  /** The GLEIF legal name of the LEI this record is keyed on. */
  gleif_legal_name?: string | null;
}

export interface EitiAssessmentBundle {
  lei: string;
  name: string;
  hq_country?: string | null;
  hq_city?: string | null;
  sectors?: string[];
  company_type?: string | null;
  business_activity?: string | null;
  match?: EitiAssessmentMatch;
  assessments?: Record<string, Record<string, EitiExpectation>>;
  subsidiaries?: EitiDeclaredSubsidiary[];
  source_snapshot?: string | null;
}

// ---------------------------------------------------------------------------
// Reading the bundle
// ---------------------------------------------------------------------------

/** The most recent assessment year present, or null.
 *
 *  Ordered by (length, value) rather than numerically, matching
 *  `sources/eiti_assessment.py::latest_year` — the two must not disagree about
 *  which year the finding sentence and the card are describing. */
export function latestAssessmentYear(bundle: EitiAssessmentBundle | null): string | null {
  const years = Object.keys(bundle?.assessments ?? {}).filter((y) => y.trim());
  if (!years.length) return null;
  return years.sort((a, b) => (a.length - b.length) || a.localeCompare(b)).at(-1) ?? null;
}

/** All assessment years, newest first. */
export function assessmentYears(bundle: EitiAssessmentBundle | null): string[] {
  return Object.keys(bundle?.assessments ?? {})
    .filter((y) => y.trim())
    .sort((a, b) => (b.length - a.length) || b.localeCompare(a));
}

export function expectation(
  bundle: EitiAssessmentBundle | null,
  shorthand: string,
  year?: string | null,
): EitiExpectation | null {
  const y = year ?? latestAssessmentYear(bundle);
  if (!y) return null;
  return bundle?.assessments?.[y]?.[shorthand] ?? null;
}

// ---------------------------------------------------------------------------
// Block 1 — the beneficial ownership disclosure assessment
// ---------------------------------------------------------------------------

/** The chip tones this card may use. Deliberately narrower than `ChipTone`:
 *  `accent` is an OpenCheck affordance and `context` is a structural
 *  observation, and neither is what an assessment result is. */
export type BoTone = "ok" | "warn" | "risk" | "neutral";

export interface BoDisclosure {
  /** The assessment year this describes. */
  year: string;
  /** EITI's verbatim result, e.g. "Expectation partially met". */
  result: string;
  /** What the company told EITI ("Yes" / "No"), when EITI recorded it. */
  response: string | null;
  tone: BoTone;
  /** The chip label — what the row asserts, in a reader's words. */
  label: string;
  /** EITI's own commentary on the result, verbatim. */
  comment: string | null;
}

/**
 * EITI's result → the tone and label the chip carries.
 *
 * **"Not available" is never a risk tone.** It is a statement about EITI's
 * assessment, not about the company, and the counter-sanctions lesson is that
 * a missing check must not be rendered as an adverse one. Anything unrecognised
 * lands in the same place for the same reason: an unknown result is not a
 * finding.
 */
const BO_DISPLAY: Record<string, { tone: BoTone; label: string }> = {
  "expectation met": { tone: "ok", label: "Discloses beneficial ownership" },
  "expectation partially met": { tone: "warn", label: "Partially discloses" },
  "expectation not met": { tone: "risk", label: "Does not disclose" },
  "not available": { tone: "neutral", label: "Not assessed" },
  "not applicable": { tone: "neutral", label: "Not assessed" },
};

export function boDisclosureTone(result: string | null | undefined): BoTone {
  return BO_DISPLAY[(result ?? "").trim().toLowerCase()]?.tone ?? "neutral";
}

export function boDisclosureLabel(result: string | null | undefined): string {
  const known = BO_DISPLAY[(result ?? "").trim().toLowerCase()];
  if (known) return known.label;
  const raw = (result ?? "").trim();
  // An unrecognised result still reports that EITI reached one, rather than
  // guessing at what it meant or silently dropping it.
  return raw ? `EITI recorded: ${raw}` : "Not assessed";
}

export function boDisclosure(bundle: EitiAssessmentBundle | null): BoDisclosure | null {
  const year = latestAssessmentYear(bundle);
  if (!year) return null;
  const exp = expectation(bundle, BO_EXPECTATION, year);
  const result = (exp?.result ?? "").trim();
  return {
    year,
    result,
    response: (exp?.response ?? "").trim() || null,
    tone: boDisclosureTone(result),
    label: boDisclosureLabel(result),
    comment: (exp?.comment ?? "").trim() || null,
  };
}

export interface DisclosureLink {
  url: string;
  /** The year whose assessment carried the link — often *not* the year on the
   *  chip. EITI populated these for 2023 and published none for 2025, so a card
   *  that showed a 2023 link under a 2025 heading would misdate the evidence. */
  year: string;
}

/** The best available link to the company's own beneficial ownership
 *  disclosure, falling back through earlier assessment years.
 *
 *  Field order matches `sources/eiti_assessment.py::bo_disclosure_url`. */
export function disclosureLink(bundle: EitiAssessmentBundle | null): DisclosureLink | null {
  for (const year of assessmentYears(bundle)) {
    const exp = bundle?.assessments?.[year]?.[BO_EXPECTATION] ?? {};
    for (const field of ["bo_url", "bo_disclosure_url", "url"] as const) {
      const value = (exp[field] ?? "").trim();
      if (value) return { url: value, year };
    }
  }
  return null;
}

/**
 * The advocacy sentence, assembled from what the row actually says.
 *
 * Every clause is conditional on evidence being present, because the point of
 * the sentence is that a *document* is not structured data — and claiming a
 * document exists where EITI published no link would be the same overstatement
 * in the other direction. Where EITI recorded a disclosure and published no
 * link, the card says exactly that rather than showing nothing.
 */
export function disclosureSentence(
  bundle: EitiAssessmentBundle | null,
): string | null {
  const bo = boDisclosure(bundle);
  if (!bo) return null;
  const name = (bundle?.name ?? "").trim() || "This company";
  const link = disclosureLink(bundle);
  const parts: string[] = [];

  const response = (bo.response ?? "").toLowerCase();
  if (response === "yes") {
    parts.push(`${name} told EITI it discloses its beneficial owners.`);
  } else if (response === "no") {
    parts.push(`${name} told EITI it does not disclose its beneficial owners.`);
  }

  const verdict: Record<string, string> = {
    "expectation met": `EITI assessed this as met for ${bo.year}.`,
    "expectation partially met": `EITI assessed this as partially met for ${bo.year}.`,
    "expectation not met": `EITI assessed this as not met for ${bo.year}.`,
  };
  const key = bo.result.toLowerCase();
  if (verdict[key]) {
    parts.push(verdict[key]);
  } else {
    parts.push(
      `EITI did not assess beneficial ownership disclosure for ${bo.year}.`,
    );
  }

  if (link) {
    parts.push(
      link.year === bo.year
        ? "The disclosure is a document, not structured data — it cannot be queried, matched, or kept current."
        : `The disclosure EITI links to was published with the ${link.year} assessment. It is a document, not structured data — it cannot be queried, matched, or kept current.`,
    );
  } else if (verdict[key]) {
    parts.push(
      `EITI records a disclosure but did not publish the link for ${bo.year}.`,
    );
  }

  return parts.join(" ");
}

// ---------------------------------------------------------------------------
// Block 3 — cross-referencing the declared names against lists we already hold
// ---------------------------------------------------------------------------

/**
 * One list to compare EITI's declared subsidiaries against.
 *
 * `available: false` is not "the list is empty" — it is "we could not ask", and
 * the two must never render the same way. A failed `/subsidiaries` fetch that
 * showed every row as "EITI only" would report a fetch failure as a finding
 * about the company, which is the Phase 124 honest-failure rule.
 */
export interface ComparisonList {
  id: string;
  /** What a reader should call this list. */
  label: string;
  /** What it measures — why a difference is not a discrepancy. */
  measures: string;
  names: string[];
  /** Identifiers this list carries, keyed by `compareKey(name)`. GLEIF children
   *  carry an LEI; B4's MEIP rows will too. Absent lists simply omit it. */
  leiByKey?: Record<string, string>;
  available: boolean;
}

export interface CrossRefRow {
  name: string;
  country: string | null;
  years: string[];
  /** ids of the comparison lists that carry this name. */
  alsoIn: string[];
  /** An LEI a comparison list holds for this name. Name-derived, so it may be
   *  linked and chipped — never asserted as an identifier, never mapped. */
  lei: string | null;
}

export interface CrossRefListSummary {
  id: string;
  label: string;
  measures: string;
  available: boolean;
  /** How many entries the comparison list itself holds. */
  listed: number;
  /** Declared names that also appear in it. */
  overlap: number;
  /** Declared names it does not carry. */
  eitiOnly: number;
  /** Its own entries that EITI does not declare — the direction that gets
   *  hidden if you only ever count from the EITI side. */
  onlyInList: number;
}

export interface CrossReference {
  declared: number;
  rows: CrossRefRow[];
  lists: CrossRefListSummary[];
}

export function crossReference(
  declared: EitiDeclaredSubsidiary[],
  lists: ComparisonList[],
): CrossReference {
  const declaredKeys = new Map<string, EitiDeclaredSubsidiary>();
  for (const sub of declared) {
    const key = compareKey(sub.name);
    if (key && !declaredKeys.has(key)) declaredKeys.set(key, sub);
  }

  const keysByList = new Map<string, Set<string>>();
  for (const list of lists) {
    keysByList.set(
      list.id,
      new Set(list.available ? list.names.map(compareKey).filter(Boolean) : []),
    );
  }

  const rows: CrossRefRow[] = declared.map((sub) => {
    const key = compareKey(sub.name);
    const alsoIn = lists
      .filter((l) => l.available && keysByList.get(l.id)?.has(key))
      .map((l) => l.id);
    const lei =
      lists
        .map((l) => (l.available ? l.leiByKey?.[key] : undefined))
        .find((v) => v) ?? null;
    return {
      name: sub.name,
      country: (sub.country ?? "").trim() || null,
      years: sub.years ?? [],
      alsoIn,
      lei,
    };
  });

  const summaries: CrossRefListSummary[] = lists.map((list) => {
    const keys = keysByList.get(list.id) ?? new Set<string>();
    const overlap = [...declaredKeys.keys()].filter((k) => keys.has(k)).length;
    return {
      id: list.id,
      label: list.label,
      measures: list.measures,
      available: list.available,
      listed: list.available ? keys.size : 0,
      overlap: list.available ? overlap : 0,
      eitiOnly: list.available ? declaredKeys.size - overlap : 0,
      onlyInList: list.available ? keys.size - overlap : 0,
    };
  });

  return { declared: declared.length, rows, lists: summaries };
}

// ---------------------------------------------------------------------------
// How the LEI on this record was arrived at
// ---------------------------------------------------------------------------

/** The builder's resolution method → a *match strength*, never a corroboration
 *  grade. Unknown methods read as the weakest, because over-stating a match is
 *  the failure that matters — `Teck` passed as a plain exact name match and was
 *  an unrelated German company. */
export function methodMatchLevel(method: string | null | undefined): "high" | "medium" | "low" {
  const m = (method ?? "").trim().toLowerCase();
  if (m === "published_lei") return "high";
  if (m === "gleif_name_exact") return "medium";
  return "low";
}

export function methodBasis(method: string | null | undefined): string {
  const m = (method ?? "").trim().toLowerCase();
  if (m === "published_lei") return "EITI publishes this LEI itself; checked by hand";
  if (m === "gleif_name_exact") return "exact name match in GLEIF, checked by hand";
  return "chosen by hand from a GLEIF shortlist";
}

// ---------------------------------------------------------------------------
// The summary tile
// ---------------------------------------------------------------------------

/** The tile above the card: the declared-subsidiary count when there is one,
 *  and the same voice when there is not — an em dash and a sentence, rather
 *  than a tile that quietly does not appear. */
export function assessmentTile(bundle: EitiAssessmentBundle | null): {
  stat: string;
  unit: string;
  sub: string;
} {
  const year = latestAssessmentYear(bundle);
  const subs = bundle?.subsidiaries ?? [];
  if (subs.length) {
    return {
      stat: subs.length.toLocaleString(),
      unit: subs.length === 1 ? "declared subsidiary" : "declared subsidiaries",
      sub: year ? `EITI Company Assessment ${year}` : "EITI Company Assessment",
    };
  }
  return {
    stat: "—",
    unit: "",
    sub: year ? `assessed ${year} · no subsidiary list filed` : "no assessment year recorded",
  };
}
